File: frontend/src/api/alerts.ts
import { apiGet, apiPatch } from "./client";
import type { AlertCenterResponse, AlertRecord } from "../types";

export function listAlerts(
  options: {
    runId?: string | null;
    status?: string | null;
    level?: string | null;
  } = {}
): Promise<AlertCenterResponse> {
  const params = new URLSearchParams();
  if (options.runId) {
    params.set("run_id", options.runId);
  }
  if (options.status) {
    params.set("status", options.status);
  }
  if (options.level) {
    params.set("level", options.level);
  }
  const query = params.toString();
  return apiGet<AlertCenterResponse>(query ? `/api/alerts?${query}` : "/api/alerts");
}

export function getAlert(alertId: string): Promise<AlertRecord> {
  return apiGet<AlertRecord>(`/api/alerts/${encodeURIComponent(alertId)}`);
}

export function acknowledgeAlert(
  alertId: string,
  acknowledgedBy?: string
): Promise<AlertRecord> {
  return apiPatch<AlertRecord>(
    `/api/alerts/${encodeURIComponent(alertId)}/acknowledge`,
    acknowledgedBy ? { acknowledged_by: acknowledgedBy } : {}
  );
}

export function resolveAlert(alertId: string): Promise<AlertRecord> {
  return apiPatch<AlertRecord>(`/api/alerts/${encodeURIComponent(alertId)}/resolve`);
}

export function ignoreAlert(alertId: string): Promise<AlertRecord> {
  return apiPatch<AlertRecord>(`/api/alerts/${encodeURIComponent(alertId)}/ignore`);
}

File: backend/app/api/alerts.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.alert import (
    AlertAcknowledgeRequest,
    AlertListResponse,
    AlertResponse,
)
from app.services.alert_service import AlertService
from app.services.event_lifecycle_service import EventLifecycleService

router = APIRouter(prefix="/api/alerts", tags=["alerts"])


@router.get("", response_model=AlertListResponse)
def list_alerts(
    run_id: str | None = Query(default=None),
    alert_status: str | None = Query(default=None, alias="status"),
    level: str | None = Query(default=None),
    db: Session = Depends(get_db),
) -> AlertListResponse:
    lifecycle = EventLifecycleService(db)
    db_alerts = lifecycle.list_alerts(
        run_id=run_id,
        status=alert_status,
        level=level,
    )
    has_db_alerts = run_id is not None and bool(lifecycle.list_alerts(run_id=run_id))
    alerts = (
        db_alerts
        if db_alerts or has_db_alerts
        else AlertService().list_alerts(
            run_id=run_id,
            status=alert_status,
            level=level,
        )
    )
    return AlertListResponse(
        alerts=[AlertResponse(**alert) for alert in alerts],
        total=len(alerts),
        run_id=run_id,
        status=alert_status,
        level=level,
    )


@router.get("/{alert_id}", response_model=AlertResponse)
def get_alert(alert_id: str, db: Session = Depends(get_db)) -> AlertResponse:
    try:
        return AlertResponse(**EventLifecycleService(db).get_alert(alert_id))
    except KeyError:
        pass
    try:
        return AlertResponse(**AlertService().get_alert(alert_id))
    except KeyError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="alert not found",
        ) from exc


@router.patch("/{alert_id}/acknowledge", response_model=AlertResponse)
def acknowledge_alert(
    alert_id: str,
    payload: AlertAcknowledgeRequest | None = None,
    db: Session = Depends(get_db),
) -> AlertResponse:
    try:
        response = AlertResponse(
            **EventLifecycleService(db).acknowledge_alert(
                alert_id,
                acknowledged_by=payload.acknowledged_by if payload else None,
            )
        )
        db.commit()
        return response
    except KeyError:
        pass
    try:
        return AlertResponse(
            **AlertService().acknowledge_alert(
                alert_id,
                acknowledged_by=payload.acknowledged_by if payload else None,
            )
        )
    except KeyError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="alert not found",
        ) from exc


@router.patch("/{alert_id}/resolve", response_model=AlertResponse)
def resolve_alert(alert_id: str, db: Session = Depends(get_db)) -> AlertResponse:
    try:
        response = AlertResponse(**EventLifecycleService(db).resolve_alert(alert_id))
        db.commit()
        return response
    except KeyError:
        pass
    try:
        return AlertResponse(**AlertService().resolve_alert(alert_id))
    except KeyError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="alert not found",
        ) from exc


@router.patch("/{alert_id}/ignore", response_model=AlertResponse)
def ignore_alert(alert_id: str, db: Session = Depends(get_db)) -> AlertResponse:
    try:
        response = AlertResponse(**EventLifecycleService(db).ignore_alert(alert_id))
        db.commit()
        return response
    except KeyError:
        pass
    try:
        return AlertResponse(**AlertService().ignore_alert(alert_id))
    except KeyError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="alert not found",
        ) from exc
